### PR TITLE
255 stateful models refactor

### PR DESCRIPTION
**Pull request by ShiriMoran**

### CODE REVIEW

**Comment by ShiriMoran** [comment]
fixes #255 (with #606)

**Review by haim-kermany — DISMISSED** [review]

File: pkg/vpcmodel/responsive_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/np-guard/models/pkg/connection"
+	"github.com/np-guard/models/pkg/netp"
+)
+
+// todo: move to analyzer
```
Comment (elazarg) [inline pkg/vpcmodel/responsive_test.go:17]:
```suggestion
```

File: pkg/vpcmodel/responsive.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/np-guard/models/pkg/netp"
+)
+
+func newTCPSet() *connection.Set {
+	return connection.TCPorUDPConnection(netp.ProtocolStringTCP, connection.MinPort, connection.MaxPort,
+		connection.MinPort, connection.MaxPort)
+}
+
+// PartitionTCPNonTCP given a connection returns its TCP and non-TCP sub-connections
+func partitionTCPNonTCP(conn *connection.Set) (tcp, nonTCP *connection.Set) {
```
Comment (adisos) [inline pkg/vpcmodel/responsive.go:19]: how are these functions related to the file name `responsive.go`? these are more general functions to be placed elsewhere?

File: pkg/vpcmodel/responsive.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	tcpFractionOfConn := newTCPSet().Intersect(conn)
+	nonTCPFractionOfConn := conn.Subtract(tcpFractionOfConn)
+	return tcpFractionOfConn, nonTCPFractionOfConn
+}
+
+// getTCPResponsiveConn returns connection object with the exact the responsive part within TCP
+// `srcToDst` represents a src-to-dst connection, and `dstToSrc` represents dst-to-src connection.
```
Comment (adisos) [inline pkg/vpcmodel/responsive.go:26]:
```suggestion
// getTCPResponsiveConn returns a connection object with the exact responsive part within TCP, given input `srcToDst` that represents a src-to-dst connection, and `dstToSrc` that represents dst-to-src connection.
```

File: pkg/vpcmodel/responsive.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+func getTCPResponsiveConn(srcToDst, dstToSrc *connection.Set) *connection.Set {
+	connTCP := srcToDst.Intersect(newTCPSet())
+	if connTCP.IsEmpty() {
+		return NoConns()
+	}
+	tcpSecondDirection := dstToSrc.Intersect(newTCPSet())
+	// flip src/dst ports before intersection
+	tcpSecondDirectionFlipped := tcpSecondDirection.SwitchSrcDstPorts()
+	// tcp connection responsive subset
+	return connTCP.Intersect(tcpSecondDirectionFlipped)
```
Comment (adisos) [inline pkg/vpcmodel/responsive.go:36]: should this function call also `detailedConnForTCPRsp` and return a `detailedConn` object instead?

File: pkg/vpcmodel/responsive_test.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/np-guard/models/pkg/connection"
+	"github.com/np-guard/models/pkg/netp"
+)
+
+func newTCPConn(t *testing.T, srcMinP, srcMaxP, dstMinP, dstMaxP int64) *connection.Set {
```
Comment (adisos) [inline pkg/vpcmodel/responsive_test.go:17]: isn't this a duplication of `newTCPSet` from `responsive.go`?

**Review by adisos — CHANGES_REQUESTED** [review]

File: pkg/vpcmodel/responsive_test.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+func newICMPconn(t *testing.T) *connection.Set {
+	t.Helper()
+	return connection.ICMPConnection(
+		connection.MinICMPType, connection.MaxICMPType,
+		connection.MinICMPCode, connection.MaxICMPCode)
+}
+
+func newTCPUDPSet(t *testing.T, p netp.ProtocolString) *connection.Set {
```
Comment (adisos) [inline pkg/vpcmodel/responsive_test.go:34]: should these functions also be placed elsewhere?

File: pkg/vpcmodel/responsive_test.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+		connection.MinPort, connection.MaxPort)
+}
+
+type responsiveTest struct {
+	name     string
+	srcToDst *connection.Set
+	dstToSrc *connection.Set
+	// expectedResponsiveConn represents the subset from srcToDst which is not related to the
+	// "non-responsive" mark (*) on the srcToDst connection,
+	expectedResponsiveConn *connection.Set
```
Comment (adisos) [inline pkg/vpcmodel/responsive_test.go:47]: consider convert to `detailedConn` instead of plain connection object

File: pkg/vpcmodel/responsive_test.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+			srcToDst:               newTCPUDPSet(t, netp.ProtocolStringTCP), // TCP all ports
+			dstToSrc:               newTCPUDPSet(t, netp.ProtocolStringTCP), // TCP all ports
+			expectedResponsiveConn: newTCPUDPSet(t, netp.ProtocolStringTCP), // TCP all ports
+		},
+		{
+			name:     "first_all_cons_second_tcp_with_ports",
+			srcToDst: connection.All(),                                              // all connections
+			dstToSrc: newTCPConn(t, 80, 80, connection.MinPort, connection.MaxPort), // TCP , src-ports: 80, dst-ports: all
+
+			// TCP src-ports: all, dst-port: 80 , union: all non-TCP conns
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive_test.go:69]: I do not understand the suffix: "also check next test objects in this file" those this refer to the comments or to something else?

File: pkg/vpcmodel/responsive.go (hunk 2 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive.go:19]: If we move these two functions to, say, `commonConnectivity.go`, then responsive.go is left with a single function. Perhaps move all to `commonConnectivity.go`?

File: pkg/vpcmodel/responsive_test.go (hunk 5 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive_test.go]: not exactly. `newTCPSet`  generates a `TCP` connection with all ports.
I moved `newTCPConn` and have `newTCPSet` use it.

File: pkg/vpcmodel/responsive_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/responsive_test.go]: refers to the comments, similar to this fix

File: pkg/vpcmodel/responsive_test.go (hunk 6 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive_test.go]: moved to `commonConnectivity`

File: pkg/vpcmodel/responsive.go (hunk 2 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive.go]: Moved to `commonConnectivity.go`, empty `responsive.go` deleted

Comment (adisos) [reply pkg/vpcmodel/responsive.go]: ok, should `commonConnectivity.go` be moved to another pkg, rather than stay in `vpcmodel` ?
maybe open a separate issue to reorg entire pkg?

File: pkg/vpcmodel/responsive.go (hunk 4 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive.go]: The inputs are `conn` (and not `detailedConn`); hence I think it makes more sense that the output will be of the same type

File: pkg/vpcmodel/responsive.go (hunk 2 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive.go]: sure. The package is too big. https://github.com/np-guard/vpc-network-config-analyzer/issues/638

File: pkg/vpcmodel/responsive_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive_test.go]: done

File: pkg/vpcmodel/responsive_test.go (hunk 7 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive_test.go:23]: Yep, makes sense. This will  strengthen the tests

File: pkg/vpcmodel/responsive.go (hunk 4 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive.go]: Indeed, in a deeper look it makes sense. I will move the function to `detailedConn.go` which also makes more sense.

File: pkg/vpcmodel/responsive_test.go (hunk 7 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/responsive_test.go]: done

**Review by adisos — APPROVED** [review]
